cpp/src/arrow/types/primitive.h

```diff
@@ -54,30 +54,54 @@ class PrimitiveArray : public Array {
-      PrimitiveArray(std::make_shared<TypeClass>(), length,     \
-          data, null_count, null_bitmap) {}                     \
-                                                                \
-  bool EqualsExact(const NAME& other) const {                   \
-    return PrimitiveArray::EqualsExact(                         \
-        *static_cast<const PrimitiveArray*>(&other));           \
-  }                                                             \
-                                                                \
-  const T* raw_data() const {                                   \
-    return reinterpret_cast<const T*>(raw_data_);               \
-  }                                                             \
-                                                                \
-  T Value(int i) const {                                        \
-    return raw_data()[i];                                       \
-  }                                                             \
+#define NUMERIC_ARRAY_DECL(NAME, TypeClass, T)                              \
+class NAME : public PrimitiveArray {                                        \
+ public:                                                                    \
+  using value_type = T;                                                     \
+  using PrimitiveArray::PrimitiveArray;                                     \
+                                                                            \
+  NAME(int32_t length, const std::shared_ptr<Buffer>& data,                 \
+      int32_t null_count = 0,                                               \
+      const std::shared_ptr<Buffer>& null_bitmap = nullptr) :               \
+      PrimitiveArray(std::make_shared<TypeClass>(), length,                 \
+          data, null_count, null_bitmap) {}                                 \
+                                                                            \
+  bool EqualsExact(const NAME& other) const {                               \
+    return PrimitiveArray::EqualsExact(                                     \
+        *static_cast<const PrimitiveArray*>(&other));                       \
+  }                                                                         \
+                                                                            \
+  std::shared_ptr<NAME> slice(int32_t start) const{                         \
+    return slice(start, length_);                                           \
+  }                                                                         \
+                                                                            \
+  std::shared_ptr<NAME> slice(int32_t start, int32_t length) const{         \
+    int32_t uFrom = start;                                                  \
+    int32_t uLength = length;                                               \
+    while(uFrom < 0 || uLength < 0) {                                       \
```

Comment:
Yes, if start is a very large negative number, the while loop will loops many, many times.
It definitely looks bad. I should chage the stategy to accept megative arguments.
I will add DCHECKs asserting respective message.
Thanks.:)